### PR TITLE
[22.05] Fix VCF handling for IGV Viewer

### LINKED ISSUE #14951: VCF display issues with IGV local
The gist:

AssertionError galaxy.datatypes.display_applications.parameters in _get_dataset_like_object
Unknown metadata name "tabix_index" provided for dataset type "vcf".

Sentry here:  https://sentry.galaxyproject.org/organizations/galaxy/issues/86713/?query=is%3Aunresolved+vcf&statsPeriod=14d

Stack:
```
AssertionError: Unknown metadata name "tabix_index" provided for dataset type "vcf".
  File "uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "starlette/middleware/base.py", line 69, in __call__
    await response(scope, receive, send)
  File "starlette/responses.py", line 260, in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
  File "anyio/_backends/_asyncio.py", line 662, in __aexit__
    raise exceptions[0]
  File "starlette/responses.py", line 256, in wrap
    await func()
  File "starlette/responses.py", line 245, in stream_response
    async for chunk in self.body_iterator:
  File "starlette/middleware/base.py", line 58, in body_stream
    raise app_exc
  File "starlette/middleware/base.py", line 36, in coro
    await self.app(scope, request.receive, send_stream.send)
  File "starlette_context/middleware/raw_middleware.py", line 96, in __call__
    await self.app(scope, receive, send_wrapper)
  File "starlette/middleware/base.py", line 69, in __call__
    await response(scope, receive, send)
  File "starlette/responses.py", line 260, in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
  File "anyio/_backends/_asyncio.py", line 662, in __aexit__
    raise exceptions[0]
  File "starlette/responses.py", line 256, in wrap
    await func()
  File "starlette/responses.py", line 245, in stream_response
    async for chunk in self.body_iterator:
  File "starlette/middleware/base.py", line 58, in body_stream
    raise app_exc
  File "starlette/middleware/base.py", line 36, in coro
    await self.app(scope, request.receive, send_stream.send)
  File "starlette/middleware/base.py", line 69, in __call__
    await response(scope, receive, send)
  File "starlette/responses.py", line 260, in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
  File "anyio/_backends/_asyncio.py", line 662, in __aexit__
    raise exceptions[0]
  File "starlette/responses.py", line 256, in wrap
    await func()
  File "starlette/responses.py", line 245, in stream_response
    async for chunk in self.body_iterator:
  File "starlette/middleware/base.py", line 58, in body_stream
    raise app_exc
  File "starlette/middleware/base.py", line 36, in coro
    await self.app(scope, request.receive, send_stream.send)
  File "starlette/exceptions.py", line 93, in __call__
    raise exc
  File "starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "starlette/routing.py", line 418, in handle
    await self.app(scope, receive, send)
  File "a2wsgi/wsgi.py", line 140, in __call__
    return await responder(scope, receive, send)
  File "a2wsgi/wsgi.py", line 179, in __call__
    raise self.exc_info[0].with_traceback(
  File "galaxy/web/framework/middleware/error.py", line 165, in __call__
    app_iter = self.application(environ, sr_checker)
  File "/cvmfs/main.galaxyproject.org/venv/lib/python3.8/site-packages/paste/recursive.py", line 85, in __call__
    return self.application(environ, start_response)
  File "galaxy/web/framework/middleware/statsd.py", line 29, in __call__
    req = self.application(environ, start_response)
  File "/cvmfs/main.galaxyproject.org/venv/lib/python3.8/site-packages/paste/httpexceptions.py", line 640, in __call__
    return self.application(environ, start_response)
  File "galaxy/web/framework/base.py", line 166, in __call__
    return self.handle_request(environ, start_response)
  File "galaxy/web/framework/base.py", line 251, in handle_request
    body = method(trans, **kwargs)
  File "galaxy/web/framework/decorators.py", line 138, in set_nocache_headers
    return func(self, trans, *args, **kwargs)
  File "galaxy/webapps/galaxy/controllers/dataset.py", line 840, in display_application
    display_link.prepare_display()
  File "galaxy/datatypes/display_applications/application.py", line 244, in prepare_display
    value = param.prepare(other_values, self.dataset_hash, self.user_hash, self.trans)
  File "galaxy/datatypes/display_applications/parameters.py", line 125, in prepare
    data = self._get_dataset_like_object(other_values)
  File "galaxy/datatypes/display_applications/parameters.py", line 105, in _get_dataset_like_object
    assert rval, f'Unknown metadata name "{self.metadata}" provided for dataset type "{data.ext}".'
    ```

### PR DESCRIPTION
**Pull request by guerler**
Fixes #14951. Will follow-up with the removal of the `display_application mako`, potentially moving the data retrieval to the API.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/datatypes/display_applications/parameters.py [hunk 1]
```diff
@@ -100,6 +100,8 @@ def _get_dataset_like_object(self, other_values):
         assert data, "Base dataset could not be found in values provided to DisplayApplicationDataParameter"
         if isinstance(data, DisplayDataValueWrapper):
             data = data.value
+        if data.state != data.states.OK:
+            return None
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/display_applications/parameters.py:104]: Are you sure this is necessary ? Both `ready` and `is_preparing` are checking the state in lib/galaxy/datatypes/display_applications/parameters.py. Changing this to None may kick off additional conversions everytime this is being polled on.

File: lib/galaxy/datatypes/display_applications/configs/igv/vcf.xml [hunk 2]
```diff
@@ -13,7 +13,7 @@
         <!-- We define url and params as normal, but values defined in dynamic_param are available by specified name -->
         <url>${redirect_url}</url>
         <param type="data" name="bgzip_file" url="galaxy_${DATASET_HASH}.vcf.gz" format="vcf_bgzip" allow_cors="true" mimetype="application/octet-stream"/>
-        <param type="data" name="tabix_file" metadata="tabix_index" url="galaxy_${DATASET_HASH}.vcf.gz.tbi" allow_cors="true" mimetype="application/octet-stream"/>
+        <param type="data" name="tabix_file" dataset="bgzip_file" metadata="tabix_index" url="galaxy_${DATASET_HASH}.vcf.gz.tbi" allow_cors="true" mimetype="application/octet-stream"/>
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/display_applications/configs/igv/vcf.xml]: This makes sense, we can skip the two-part conversion to bgzip and tabix, but this probably also hides a bug with finding the converted dataset.

Comment (mvdbeek) [reply lib/galaxy/datatypes/display_applications/configs/igv/vcf.xml]: I wonder if this ever worked for igv, the other display applications specify `format="tabix"` for displaying vcfs, and I assume that still works. We should switch them all over, I can write a test for that.

File: lib/galaxy/datatypes/display_applications/parameters.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/datatypes/display_applications/parameters.py:104]: Yep, that breaks adding a new vcf

Comment (mvdbeek) [reply lib/galaxy/datatypes/display_applications/parameters.py:104]: nevermind, got the wrong x-accel-redirect set up 😓

Comment (guerler) [reply lib/galaxy/datatypes/display_applications/parameters.py:104]: No worries, thanks for looking into it.

File: lib/galaxy/datatypes/display_applications/configs/igv/vcf.xml (hunk 2 above)
Comment (guerler) [reply lib/galaxy/datatypes/display_applications/configs/igv/vcf.xml]: I was wondering that too and I doubt that it did work before, and I agree accessing the `metadata` of an already converted dataset is probably better than having to further convert it.

**Comment by mvdbeek** [comment]
> potentially moving the data retrieval to the API.

I think that's a good idea, but we should first work out a way to generalize the data access permissions (the scheme here is encoding the user id with the dataset create time as the secret -> https://github.com/guerler/galaxy/blob/55080e719fe501688154a8dd2945baff4c90a649/lib/galaxy/datatypes/display_applications/util.py#L16-L30). I don't think we want to make this part of the "official" API. For the time being I'd keep this as a controller. Once we have figured out how we want to create and store access token we can just redirect straight to the regular datasets API.

**Review by mvdbeek — APPROVED** [review]

**Pull request merged**